djstripe/stripe_objects.py

```diff
@@ -1097,6 +1187,61 @@ def _stripe_object_to_charge(cls, target_cls, data):
         if "charge" in data and data["charge"]:
             return target_cls._get_or_create_from_stripe_object(data, "charge")[0]
 
+    @classmethod
+    def upcoming(cls, api_key=settings.STRIPE_SECRET_KEY, **kwargs):
+        """
+        Gets the upcoming preview invoice (singular) for a customer.
+
+        As per the Stripe docs: "At any time, you can preview the upcoming
+        invoice for a customer. This will show you all the charges that are
+        pending, including subscription renewal charges, invoice item charges,
+        etc. It will also show you any discount that is applicable to the
+        customer."
+
+        See for details: https://stripe.com/docs/api#upcoming_invoice
+
+        :param customer: The identifier of the customer whose upcoming invoice
+        you'd like to retrieve.
+        :param coupon: The code of the coupon to apply.
+        :param subscription: The identifier of the subscription to retrieve an
+        invoice for.
+        :param subscription_plan: If set, the invoice returned will preview
+        updating the subscription given to this plan, or creating a new
+        subscription to this plan if no subscription is given.
+        :param subscription_prorate: If previewing an update to a subscription,
+        this decides whether the preview will show the result of applying
+        prorations or not.
+        :param subscription_proration_date: f previewing an update to a
+        subscription, and doing proration, subscription_proration_date forces
+        the proration to be calculated as though the update was done at the
+        specified time.
+        :param subscription_quantity: If provided, the invoice returned will
+        preview updating or creating a subscription with that quantity.
+        :param subscription_trial_end: If provided, the invoice returned will
+        preview updating or creating a subscription with that trial end.
+        :returns: The upcoming preview invoice.
+        :rtype: ``djstripe.models.UpcomingInvoice``
+        """
+        from . models import Invoice, UpcomingInvoice
+        if not issubclass(cls, UpcomingInvoice):
+            if cls is Invoice:
+                cls = UpcomingInvoice
+            else:
+                assert False, (
+                    "Class argument needs to be derived from UpcomingInvoice")
+
+        try:
+            data = cls._api().upcoming(api_key=api_key, **kwargs)
+        except InvalidRequestError as ex:
+            if str(ex) != "Nothing to invoice for customer":
+                six.reraise(*sys.exc_info())
```

Comment:
It basically means that the stack trace will continue up into the library rather than stopping at the point of where the exception is raised again manually.  This will only happen where it is supported and probably requires the exception chaining support that python3.x has (as per [PEP-3134](https://www.python.org/dev/peps/pep-3134/)).

Here's what the six documentation says for reraise:

> Reraise an exception, possibly with a different traceback. **In the simple case, `reraise(*sys.exc_info())` with an active exception (in an except block) reraises the current exception with the last traceback.** A different traceback can be specified with the exc_traceback parameter. Note that since the exception reraising is done within the reraise() function, Python will attach the call frame of reraise() to whatever traceback is raised.